### PR TITLE
Implement void[]-based ring queue in Mirror handler

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
Fixes two issues:
1. We encountered segfaults under high load. Investigation revealed that
   this is due to casting a `void[]*` to a pointer to another type of array
   (as we were doing in the old queue implementation). The exact cause of
   the segfaults was not discovered, but it appears to be related to GC
   corruption.
   See https://github.com/sociomantic-tsunami/dhtnode/issues/15.
2. The previous queue implementation was pretty inefficient: it required a
   memory shift of the entire queue after each pop. The new implementation
   does not require any memory shifting.

### CODE REVIEW

File: src/dhtproto/node/neo/request/Mirror.d [hunk 1]
```diff
@@ -260,7 +371,8 @@ public abstract scope class MirrorProtocol_v0
                 (the request code and version are assumed to be extracted)
 
     ***************************************************************************/
-
+import ocean.io.Stdout;
```
Comment (gavin-norman-sociomantic) [inline src/dhtproto/node/neo/request/Mirror.d]: Haha, ok I'll remove this debug stuff ;)

File: src/dhtproto/node/neo/request/Mirror.d [hunk 2]
```diff
@@ -65,18 +69,50 @@ public abstract scope class MirrorProtocol_v0
...
+            Params:
+                buf = array to use to back queue storage
+
+        ***********************************************************************/
+
+        public void initialise ( void[]* buf )
+        {
+            this.max_elems = (max_size / E.sizeof) * E.sizeof;
+            buf.length = this.max_elems * E.sizeof;
+            this.queue = (cast(E*)buf.ptr)[0 .. this.max_elems];
```
Comment (david-eckardt-sociomantic) [inline src/dhtproto/node/neo/request/Mirror.d]: Suggest
- `ref void[] buf` now that you don't store a pointer to the array any more,
- `(cast(E[])*buf)[0 .. this.max_elems]` (without the `*` for `ref void[] buf`).

Comment (david-eckardt-sociomantic) [reply src/dhtproto/node/neo/request/Mirror.d]: Actually `this.queue = cast(E[])buf;` should be sufficient

Comment (gavin-norman-sociomantic) [reply src/dhtproto/node/neo/request/Mirror.d]: Updated.

**Review by david-eckardt-sociomantic — APPROVED** [review]

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by joseph-wakeling-sociomantic** [comment]
Can confirm that this (in combination with swarm changes in v4.2.4) results in no segfaults during a mirror request for my use-case.

**Pull request merged**